### PR TITLE
raftclient: count term and index in estimated size

### LINKED ISSUE #9714: batch messages can be too large in new raft client implementation
## Bug Report

New raft client implementation tends to batch more messages than before and we found it exceeded the max grpc messages in internal tests.

```
[2021/02/25 17:42:42.658 +08:00] [ERROR] [raft_client.rs:431] ["connection aborted"] [addr=] [receiver_err="Some(RpcFailure(RpcStatus { stat
us: 8-RESOURCE_EXHAUSTED, details: Some(\"Sent message larger than max (10705556 vs. 10485760)\") }))"] [sink_error=Some(RemoteStopped)] [store_id=7]
```

Calculating the message size using `Message::compute_size` should make it correct at the cost with extra CPU usage. We may also check what fields can be missed as even max batch size (128) is reached, there was about 1K extra size per message.

### PR DESCRIPTION
**Pull request by BusyJay**
### What problem does this PR solve?

Issue Number: close #9714.

Problem Summary:

If there are many entries in a message, the estimated size of message can be way smaller than
the actual size. This PR fixes the error by also counting index and term in estimation.

It also remove the hard limit as the estimation is closed enough.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
Fix connection abort when too many raft entries are batched into one messages
```

### CODE REVIEW

File: src/server/raft_client.rs [hunk 1]
```diff
@@ -174,7 +174,9 @@ impl Buffer for BatchMessageBuffer {
         let mut msg_size = msg.start_key.len()
             + msg.end_key.len()
             + msg.get_message().context.len()
-            + msg.extra_ctx.len();
+            + msg.extra_ctx.len()
+            // index: 3, term: 2, data tag and size: 3, entry tag and size: 3
+            + 11 * msg.get_message().get_entries().len();
```
Comment (NingLin-P) [inline src/server/raft_client.rs:179]: There are also other fields in `eraftpb::Message` like `to`, `from`, `term`, etc.

Comment (BusyJay) [reply src/server/raft_client.rs:179]: Yes, but there are at most 128 messages, so the size can be omitted.

File: src/server/raft_client.rs [hunk 2]
```diff
@@ -568,7 +570,6 @@ where
 
         let cb = ChannelBuilder::new(self.builder.env.clone())
             .stream_initial_window_size(self.builder.cfg.grpc_stream_initial_window_size.0 as i32)
-            .max_send_message_len(self.builder.cfg.max_grpc_send_msg_len)
```
Comment (tonyxuqqi) [inline src/server/raft_client.rs:571]: So the max message limit is removed in grpc layer,  do we have metrics ( histogram) of the grpc size for this particular Raft batch API?  This is to validate if the size check in RaftClient code works in all the cases and help to diagnostics perf related issues

File: src/server/raft_client.rs (hunk 1 above)
Comment (tonyxuqqi) [inline src/server/raft_client.rs:179]: minor:  per entry length can be added in the loop below.
msg_size += entry.data.len() + 11; // index: 3, term: 2, data tag and size: 3, entry tag and size: 3

File: src/server/raft_client.rs (hunk 2 above)
Comment (BusyJay) [reply src/server/raft_client.rs:571]: Not yet. I plan to add one once middleware is supported in grpc-rs.

File: src/server/raft_client.rs (hunk 1 above)
Comment (BusyJay) [reply src/server/raft_client.rs:179]: Yes, but it's efficient to use multiply than loop.

File: src/server/raft_client.rs (hunk 2 above)
Comment (hicqu) [reply src/server/raft_client.rs:571]: Can we remove the entry `max_grpc_send_msg_len` from `Config`? Or mark it as deprecated.

Comment (BusyJay) [reply src/server/raft_client.rs:571]: No, it's used for controlling the batch size.

**Review by NingLin-P — APPROVED** [review]

**Review by hicqu — APPROVED** [review]

**Pull request merged**